File: src/RNA2D.cpp
#include "../include/RNA2D.h"


std::string only_paired(std::string dot_bracket)
{
    // removes the "." characters from the dot_bracket
    assert(is_valid_dot_bracket(dot_bracket));  // only apply on legal dot_bracket
    std::string ret = std::string(dot_bracket);
    ret.erase(std::remove(ret.begin(), ret.end(), '.'), ret.end());
    return ret;
}


bool is_valid_dot_bracket(std::string dot_bracket)
{
    // tests Vienna dot-bracket for illegal structure (or symbol)
    int counter = 0;

    for (unsigned int i = 0; i < dot_bracket.size(); ++i)
    {
        char c = dot_bracket[i];
        if (c == '(')               // stack
        {
            counter += 1;
        }
        else if (c == ')')          // unstack
        {
            counter -= 1;
        }
        else if (c != '.')          // illegal character
        {
            return false;
        }
        if (counter < 0)            // left unbalanced
        {
            return false;
        }
    }
    if (counter != 0)               // right unbalanced
    {
        return false;
    }
    else                            // correct dotbracket
    {
        return true;
    }
}

File: mccons2/src/rna/RNA2D.java
package rna;


import java.util.ArrayList;

public class RNA2D {
    private final String dotBracket;
    private final String sequence;


    //region CONSTRUCTORS

    public RNA2D(String dotBracket, String sequence) {

        assert isValid2DForest(dotBracket);
        assert isValidRNASequence(sequence);

        this.dotBracket = dotBracket;
        this.sequence = sequence;
    }

    public RNA2D(RNA2D rna)
    {
        this.dotBracket = rna.getDotBracket();
        this.sequence = rna.getSequence();
    }
    //endregion


    //region INPUT CHECKERS
    /**
     * verifies that the arguments given to the constructor build a legal tree
     * @param dotBracket string representation, usually Vienna dot bracket
     * @return true if a valid tree is build from the arguments
     */
    public static boolean isValid2DForest(String dotBracket)
    {
        int counter = 0;
        for (char c : dotBracket.toCharArray())
        {
            if (c == '(')        // stack
            {
                counter += 1;
            }
            else if (c == ')')   // unstack
            {
                counter -= 1;
            }
            else if (c != '.')   // illegal character
            {
                return false;
            }
            if (counter < 0)     // left unbalanced
            {
                return false;
            }
        }
        return counter == 0;
    }


    /**
     * verifies if the RNA sequence given is composed of AUGC
     * @param sequence RNA sequence
     * @return true if it is a valid sequence of AUGC, false otherwise
     */
    public static boolean isValidRNASequence(String sequence)
    {
        boolean isValid = true;
        for (char c : sequence.toUpperCase().toCharArray())
        {
            if (!(c == 'A' || c =='U' || c == 'G' || c == 'C'))
            {
                isValid = false;
                break;
            }
        }
        return isValid;
    }
    //endregion


    //region ACCESS

    public String getDotBracket() {
        return new String(dotBracket);
    }

    public String getSequence() {
        return new String(sequence);
    }

    //endregion


    //region COMPARISON
    @Override
    public boolean equals(Object other)
    {
        boolean result = false;
        if (other instanceof RNA2D)
        {
            RNA2D that = (RNA2D) other;
            result = sequence.compareTo(that.getSequence()) == 0 && dotBracket.compareTo(that.getDotBracket()) == 0;
        }
        return result;
    }
    //endregion


}
